File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int res = 0;
public:
    int goodNodes(TreeNode* root) {
        dfs(root, root -> val);

        return res;
    }

    void dfs(TreeNode* node, int maxValue) {
        if (!node) {
            return;
        }

        res += (node -> val >= maxValue) ? 1 : 0;
        maxValue = max(node -> val, maxValue);

        dfs(node -> left, maxValue);
        dfs(node -> right, maxValue);
    }
};

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def DFS(root: TreeNode, maxValue: int) -> int:
            if not root:
                return 0
            
            good = 1 if root.val >= maxValue else 0
            maxValue = max(root.val, maxValue)

            return good + DFS(root.left, maxValue) + DFS(root.right, maxValue)
        
        return DFS(root, root.val)
        

        